datalad/cmd2.py

```diff
@@ -0,0 +1,86 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Wrapper for command calls, allowing for dry runs and output handling
+
+"""
+
+
+import subprocess
+import shlex
+import sys
+import logging
+
+lgr = logging.getLogger('datalad.cmd')
+
+
+class Runner(object):
+    """Provides a wrapper for system calls.
+
+    An object of this class provides a method to make system calls.
+    Outputs (stdout and stderr) are streamed to stdout during execution as if you were calling it from command line.
+    Additionally allows for dry runs. This is achieved by initializing the `Runner` with `dry=True`.
+    The Runner will then collect all calls as strings in `cmdBuffer`.
+    """
+
+    __slots__ = ['cmdBuffer', 'dry']
+
+    def __init__(self, dry=False):
+        self.dry = dry
+        self.cmdBuffer = []
+
+    def run(self, cmd):
+        """
+
+        Parameters
+        ----------
+        cmd : str
+          String defining the command call.
+
+        Returns
+        -------
+        Status code as returned by the called command.
+        """
+
+        self.log("Running: %s" % cmd)
+
+        if not self.dry:
+            # pipeArgs = shlex.split(cmd)
+
+            proc = subprocess.Popen(cmd, stdout=sys.stdout, stderr=subprocess.STDOUT, shell=True)
```

Comment:
> anyways -- how does this improve upon existing solution??
The intended difference is the use of subprocess and live output of stdout as discussed.

> I see no drycall which allowed for dry execution/reporting of python native functions
I would have preserved it, when replacing the old file. But how do you want to use it?

> I see no logging of output at all (which was previously nicely indented with '|')
See the comment before

> getting back -- was probably better just to fix up existing cmd.py to use subprocess.Popen
> instead
From my point of view, this is what I'm doing ;-)